### PR TITLE
Tests added for BooService, and BookRepository

### PR DESCRIPTION
**Pull request by VKhmil**

### CODE REVIEW

**Comment by VKhmil** [comment]
image

File: src/test/java/com/bookappstore/controller/BookControllerTest.java [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+                    connection,
+                    new ClassPathResource(
+                            "database/test/categories/add-categories-to-table.sql")
+            );
+            ScriptUtils.executeSqlScript(
+                    connection,
+                    new ClassPathResource(
+                            "database/test/books/add-books-and-categories-into-table.sql")
+            );
+        }
```
Comment (liliia-ponomarenko) [inline src/test/java/com/bookappstore/controller/BookControllerTest.java]: You can use annotation SQL below the class

File: src/test/java/com/bookappstore/controller/BookControllerTest.java [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+
+    @AfterAll
+    static void afterAll(
+            @Autowired DataSource dataSource
+    ) {
+        teardown(dataSource);
+    }
+
+    @SneakyThrows
+    private static void teardown(DataSource dataSource) {
```
Comment (liliia-ponomarenko) [inline src/test/java/com/bookappstore/controller/BookControllerTest.java]: the same

File: src/test/java/com/bookappstore/controller/BookControllerTest.java [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+    @Sql(scripts = {
+            "classpath:database/test/books/add-one-book.sql",
+    },
+            executionPhase = Sql.ExecutionPhase.BEFORE_TEST_METHOD)
+    @Sql(scripts = {
+            "classpath:database/test/books/delete-book-from-books_categories.sql",
+            "classpath:database/test/books/remove-book-from-table-books.sql",
+    },
+            executionPhase = Sql.ExecutionPhase.AFTER_TEST_METHOD)
+    void updateByValidId_ShouldReturnUpdateBook() throws Exception {
```
Comment (liliia-ponomarenko) [inline src/test/java/com/bookappstore/controller/BookControllerTest.java]:
```suggestion
    void updateByValidId_ShouldReturnUpdateBook_Ok() throws Exception {
```

File: src/test/java/com/bookappstore/controller/CategoryControllerTest.java [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+                .andExpect(status().isOk())
+                .andReturn();
+
+        CategoryDto actualDto = objectMapper.readValue(result.getResponse()
+                .getContentAsString(),
+                CategoryDto.class);
+        Assertions.assertEquals(updateCategoryDto.getName(), actualDto.getName());
+        Assertions.assertEquals(updateCategoryDto.getDescription(), actualDto.getDescription());
+    }
+
```
Comment (liliia-ponomarenko) [inline src/test/java/com/bookappstore/controller/CategoryControllerTest.java]:
```suggestion
```

File: src/test/java/com/bookappstore/repository/BookRepositoryTest.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+            executionPhase = Sql.ExecutionPhase.BEFORE_TEST_METHOD)
+    @Sql(scripts = {
+            "classpath:database/test/books/delete-books-categories.sql",
+            "classpath:database/test/books/remove-books-from-table-books.sql",
+            "classpath:database/test/categories/delete-categories.sql",
+    },
+            executionPhase = Sql.ExecutionPhase.AFTER_TEST_METHOD)
+    @DirtiesContext(classMode = DirtiesContext.ClassMode.AFTER_EACH_TEST_METHOD)
+    public void findAllByCategoryId_ExistedId_Success() {
+        Category category = categoryRepository.findById(12L).get();
```
Comment (liliia-ponomarenko) [inline src/test/java/com/bookappstore/repository/BookRepositoryTest.java]: Don't call any other repository methods in test case

File: src/test/java/com/bookappstore/service/impl/BookServiceImpTest.java [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+
+        assertNotNull(result);
+        assertEquals(bookDtos.size(), result.size());
+
+        verify(bookSpecificationBuilder).build(searchParams);
+        verify(bookRepository).findAll(specification);
+        verify(bookMapper).toDtoList(books);
+    }
+
+    private Category createCategory(Long id, String name) {
```
Comment (liliia-ponomarenko) [inline src/test/java/com/bookappstore/service/impl/BookServiceImpTest.java:208]: I really like that in this class you put the creation of test objects into private methods, you can apply this to other test classes.

File: src/test/java/com/bookappstore/service/impl/BookServiceImpTest.java [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+        book.setPrice(BigDecimal.valueOf(50));
+        book.setDescription("Test description");
+        book.setIsbn("978-3-16-148410-0");
+        book.setCoverImage("http://example.com/cover.jpg");
+        return book;
+    }
+
+    private Book createTestBookWithoutId() {
+        return createTestBook();
+    }
```
Comment (liliia-ponomarenko) [inline src/test/java/com/bookappstore/service/impl/BookServiceImpTest.java]: Do you need this method that just calls another method?

File: src/test/java/com/bookappstore/service/impl/BookServiceImpTest.java [hunk 8]
```diff
@@ -0,0 +1,282 @@
...
+
+        assertThat(bookDtoServiceById).isEqualTo(bookDto);
+        verify(bookRepository, times(1)).findBookById(idBook);
+        verify(bookMapper, times(1)).toDto(book);
+        verifyNoMoreInteractions(bookRepository, bookMapper);
+    }
+
+    @Test
+    @DisplayName("Using wrong id, method must throw an exception")
+    void getById_NonExistingId_ShouldReturnException() {
```
Comment (liliia-ponomarenko) [inline src/test/java/com/bookappstore/service/impl/BookServiceImpTest.java]:
```suggestion
    void getById_NonExistingId_ShouldThrowsException() {
```

File: src/test/java/com/bookappstore/service/impl/BookServiceImpTest.java [hunk 9]
```diff
@@ -0,0 +1,282 @@
...
+        when(bookRepository.findBookById(anyLong())).thenReturn(Optional.empty());
+        Assertions.assertThrows(EntityNotFoundException.class,
+                () -> bookService.findById(anyLong()));
+    }
+
+    @Test
+    @DisplayName("Saving book with correct fields in a right way")
+    void save_NormalBook_ShouldReturnBook() {
```
Comment (liliia-ponomarenko) [inline src/test/java/com/bookappstore/service/impl/BookServiceImpTest.java]:
```suggestion
    void save_ValidlBook_ShouldReturnBook() {
```

File: src/test/java/com/bookappstore/service/impl/BookServiceImpTest.java [hunk 10]
```diff
@@ -0,0 +1,282 @@
...
+        when(bookRepository.save(bookWithoutId)).thenReturn(bookWithoutId);
+        when(bookMapper.toDto(bookWithoutId)).thenReturn(bookDto);
+
+        BookDto bookServiceById = bookService.save(testBookRequestDto);
+
+        assertThat(bookServiceById).isEqualTo(bookDto);
+    }
+
+    @Test
+    public void testFindAllBooksWithCategories() {
```
Comment (liliia-ponomarenko) [inline src/test/java/com/bookappstore/service/impl/BookServiceImpTest.java]: use the similar naming style for test cases ;)

File: src/test/java/com/bookappstore/service/impl/BookServiceImpTest.java [hunk 11]
```diff
@@ -0,0 +1,282 @@
...
+    }
+
+    @Test
+    @DisplayName("Update book by Id")
+    public void updateBook_UpdateBookById_Correct() {
+        Set<Category> categories = Set.of(createCategory(1L, "Fiction"));
+        Book book = createTestBook();
+        CreateBookRequestDto createBookRequestDto = createTestBookRequestDto();
+        Mockito.when(categoryRepository.findByIdIn(createBookRequestDto.getCategoriesIds()))
+                .thenReturn(categories);
```
Comment (liliia-ponomarenko) [inline src/test/java/com/bookappstore/service/impl/BookServiceImpTest.java:141]: Keep all mocks in one group, good practice to follow GWT grouping as well
// Given
... 
// When
...
// Than
...

File: src/test/java/com/bookappstore/service/impl/CategoryServiceImplTest.java [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+import org.junit.jupiter.api.Test;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.PageImpl;
+import org.springframework.data.domain.PageRequest;
+import org.springframework.data.domain.Pageable;
+
+class CategoryServiceImplTest {
```
Comment (liliia-ponomarenko) [inline src/test/java/com/bookappstore/service/impl/CategoryServiceImplTest.java:29]:
```suggestion
@ExtendWith(MockitoExtension.class)
class CategoryServiceImplTest {
```
missed

File: src/test/java/com/bookappstore/service/impl/CategoryServiceImplTest.java [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+    private CategoryServiceImpl categoryService;
+
+    private Category category;
+    private CategoryDto categoryDto;
+
+    @BeforeEach
+    void setUp() {
+        MockitoAnnotations.openMocks(this);
+
+        // Ініціалізація тестових даних
```
Comment (liliia-ponomarenko) [inline src/test/java/com/bookappstore/service/impl/CategoryServiceImplTest.java]:
```suggestion
```

File: src/test/java/com/bookappstore/service/impl/CategoryServiceImplTest.java [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+        category.setName("Test Category");
+        category.setDescription("Test Description");
+
+        categoryDto = new CategoryDto();
+        categoryDto.setName("Test Category");
+        categoryDto.setDescription("Test Description");
+    }
+
+    @Test
+    void testSave() {
```
Comment (liliia-ponomarenko) [inline src/test/java/com/bookappstore/service/impl/CategoryServiceImplTest.java]: let's fix test names in this class ;)

File: src/test/resources/database/test/books/add-books-and-categories-into-table.sql [hunk 15]
```diff
@@ -0,0 +1,2 @@
+INSERT INTO books_categories (category_id, book_id)
+VALUES (12, 12), (12, 13);
```
Comment (liliia-ponomarenko) [inline src/test/resources/database/test/books/add-books-and-categories-into-table.sql]: Add one empty line at the end of the file (check other files)

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Great 🔥 Let's improve your solution ;)

**Review by okuzan — APPROVED** [review]
Good job!

**Pull request merged**